### PR TITLE
feat: add `tip` stdcheat

### PR DESCRIPTION
**Pull request by ZeroEkkusu**
Gm! This is a new `tip` stdcheat.

### Motivation
Simplify funding an account with a set amount of an arbitrary ERC20 token.

### Solution
A convenient solution you can use in tests:
```solidity
tip(alice, 10000e18, dai);

// Let's test it
assertEq(daiToken.balanceOf(alice), 10000e18); // [PASS]
```
Especially useful when forking.

Edit: Need to make a few changes to make it work with ~~every~~ a majority of tokens. **✔️ DONE**

### CODE REVIEW

**Comment by ZeroEkkusu** [comment]
# Update

- [EIP-20: Token Standard](https://eips.ethereum.org/EIPS/eip-20) requires only that the token has the [`balanceOf`](https://eips.ethereum.org/EIPS/eip-20#balanceof) method, not how to store balances or report them.
- We can only assume that the token will report a balance honestly. Also, different tokens store balances in different ways. For example, DAI stores them in `balanceOf`, 1INCH in `_balances`, MATIC in `balances`.

## Solution

- A majority of tokens read a balance  immediately when the `balanceOf` method is called, but we must not assume this to always be the case.
- I have changed how the `tip` stdcheat works, so it can always set a balance to what we want, when the balance is read immediately after calling `balanceOf`. If `balanceOf` reads some other values before reading a balance, you must specify the name of the mapping that holds balances:
```solidity
// WHEN READ IMMEDIATELY:

/// function balanceOf(address account) public view override returns (uint256) {
///     return _balances[account];
/// }

tip(address(1inch), 10000e18, alice);


// WHEN NOT READ IMMEDIATELY:

/// function balanceOf(address account) public view override returns (uint256) {
///     bytes32 something = someStateVariable;
///     ...
///     return _balances[account];
/// }

tip(address(atypicalToken), "_balances", 10000e18, alice);
```

 # Ready for review!

**Comment by brockelmore** [comment]
is this ready for review now?

**Comment by ZeroEkkusu** [comment]
> is this ready for review now?

Yes, it is.

File: src/stdlib.sol [hunk 1]
```diff
@@ -62,6 +62,18 @@ abstract contract stdCheats {
         vm_std_cheats.startPrank(who, origin);
     }
 
+    // Allows you to the balance of an account for a majority of tokens
+    // Be careful not to break something!
+    function tip(address to, uint256 give, address which) public {
```
Comment (brockelmore) [inline src/stdlib.sol:67]: Why dont you just use stdstore instead? i imagine there will be more useful things like this that pop up

Comment (brockelmore) [reply src/stdlib.sol:67]: can we change the variable arrangement as well? i think (token, to, give) is most intuitive

**Review by brockelmore — CHANGES_REQUESTED** [review]

File: src/stdlib.sol (hunk 1 above)
Comment (brockelmore) [reply src/stdlib.sol]: https://mirror.xyz/brocke.eth/PnX7oAcU4LJCxcoICiaDhq_MUUu9euaM8Y5r465Rd2U < see here. I use stdstore

Comment (ZeroEkkusu) [reply src/stdlib.sol]: thanks, brock. i didn't know u can use it like that.

the article was informative!

Comment (ZeroEkkusu) [reply src/stdlib.sol]: it's ready, ser @brockelmore

**Review by brockelmore — APPROVED** [review]
lgtm!

**Comment by shuklaayush** [comment]
Hey
Thought I'd just comment here instead of creating a new issue. Isn't the term "tip" a bit misleading since the cheat code is setting the balance and not adding to it? Maybe an overloaded `deal` would be better. Also, might make sense to update `totalSupply` along with the balance

**Comment by ZeroEkkusu** [comment]
@shuklaayush

I had considered making `tip` add to a balance, but setting a balance is more useful. Do you have a better name for it? I don't want hevm cheatcodes and stdcheats to be confused, which is why it is not called `deal`.

Since this stdcheat takes the address of an *arbitrary* ERC20 token, we aren't aware how the token operates. For example, it's `totalSupply` might be essential for some internal logic and changing it could break something, which is why I went with a minimal approach.

**Comment by shuklaayush** [comment]
Ah, cool. Didn't realize that you intentionally avoided naming it `deal`. Only alternative I can come up with off the top of my head is `allot` but I guess `tip` is fine too

While I agree that it's better to keep the logic minimal, I think practically it's more likely that someone will run into an error when `totalSupply` isn't equal to the sum of balances as compared to an error where the token contract implements some special logic around `totalSupply`

Anyways, I don't have strong opinions on either. Thought I'd share my two cents since I did some similar stuff when testing out contracts with foundry

**Pull request merged**